### PR TITLE
fix(#1943): new component filterchip

### PR DESCRIPTION
**Pull request by syedszeeshan**
A separate PR for the documentation update is also needed.

# Before (the change)

# After (the change)

## Make sure that you've checked the boxes below before you submit the PR

- [x] I have read and followed the [setup steps](https://goa-dio.atlassian.net/wiki/spaces/DS/pages/3086385489/Contribution+-+web+content#%5BinlineExtension%5DCode-contributions)
- [x] I have created necessary unit tests
- [x] I have tested the functionality in both React and Angular.

## Steps needed to test

**Playground Angular Example**
Typescript
```
import { Component } from "@angular/core";

@Component({
  selector: "abgov-chip",
  templateUrl: "./filter-chip.component.html",
  styleUrl: "./filter-chip.component.css",
})
export class FilterChipComponent {
  chips: string[] = ["Chip 1", "Chip 2", "Chip 3"];
  activeFilters: string[] = [];
  typedChips: string[] = ["Typed Chip 1", "Typed Chip 2", "Typed Chip 3"];
  inputValue = "";

  deleteChip(chip: string): void {
    this.chips = this.chips.filter((c) => c !== chip);
  }

  onClick(): void {
    console.log("Chip clicked");
  }
```

Html
```
<h1>Filter Chip</h1>

<h2>Delete event</h2>
<goa-filter-chip
  *ngFor="let chip of chips"
  [content]="chip"
  (_click)="deleteChip(chip)"
  mr="s"
></goa-filter-chip>

<h2>Error state</h2>
<goa-filter-chip content="Filter Chip Text" error="true"  mr="s"></goa-filter-chip>
<goa-filter-chip
  content="Filter Chip Text"
  error="true"
  (_click)="onClick()"
></goa-filter-chip>


<h2>Unfilled Close Icon (Default)</h2>
<goa-filter-chip content="This is a really long chip content that wraps when the text gets too long"></goa-filter-chip>

<h2>Filled Close Icon (On Hover)</h2>
<p>Hover over the chip to see the close icon change:</p>
<goa-filter-chip content="Hover Me"></goa-filter-chip>

<h2>Multiple Filter Chips</h2>
<div>
  <goa-filter-chip content="Category 1" (_click)="onFilterRemove('Category 1')"  mr="s"></goa-filter-chip>
  <goa-filter-chip content="Category 2" (_click)="onFilterRemove('Category 2')"  mr="s"></goa-filter-chip>
  <goa-filter-chip content="Category 3" (_click)="onFilterRemove('Category 3')"></goa-filter-chip>
</div>

```

**Playground React Code**

```
import React, { useState, useRef, KeyboardEvent, ChangeEvent } from "react";
import { GoAButton, GoAFilterChip, GoAInput } from "@abgov/react-components";

export default function FilterChipComponent() {
  const [chips, setChips] = useState(["Chip 1", "Chip 2", "Chip 3"]);
  const [activeFilters, setActiveFilters] = useState<string[]>([]);

  const deleteChip = (chip: string) => {
    setChips((prevChips) => prevChips.filter((c) => c !== chip));
  };

  const onFilterRemove = (category: string) => {
    console.log(`Filter removed: ${category}`);
  };

  return (
    <div style={{ marginTop: "2rem" }}>
      <h1 className="text-2xl font-bold mb-4">GoA Filter Chip Playground</h1>

      <section style={{ marginTop: "2rem" }}>
        <h2>Delete event</h2>
        <div className="flex flex-wrap gap-2">
          {chips.map((chip) => (
            <GoAFilterChip
              key={chip}
              content={chip}
              onClick={() => deleteChip(chip)}
              mr="s"
            />
          ))}
        </div>
      </section>

      <section style={{ marginTop: "2rem" }}>
        <h2>Error state</h2>
        <GoAFilterChip content="Filter Chip Text" error={true} mr="s" />
        <GoAFilterChip
          content="Filter Chip Text"
          error={true}
          onClick={() => console.log("Chip clicked")}
        />
      </section>

      <section style={{ marginTop: "2rem" }}>
        <h2>Unfilled Close Icon (Default)</h2>
        <GoAFilterChip content="Hover Me" />
      </section>

      <section style={{ marginTop: "2rem" }}>
        <h2>Filled Close Icon (On Hover)</h2>
        <p>Hover over the chip to see the close icon change:</p>
        <GoAFilterChip content="Hover Me" />
      </section>

      <section style={{ marginTop: "2rem" }}>
        <h2>Multiple Filter Chips</h2>
        <div className="flex flex-wrap gap-2">
          <GoAFilterChip
            content="Category 1"
            onClick={() => onFilterRemove("Category 1")}
            mr="s"
          />
          <GoAFilterChip
            content="Category 2"
            onClick={() => onFilterRemove("Category 2")}
            mr="s"
          />
          <GoAFilterChip
            content="Category 3"
            onClick={() => onFilterRemove("Category 3")}
          />
        </div>
      </section>
    </div>
  );
}

```

### CODE REVIEW

**Comment by lizhuomeng71** [comment]
1） Figma File indicate the border color should be 666666  (--goa-color-greyscale-700). the actual boarder color 949494 (--goa-color-greyscale-500)
2） Figma File indicate the text line height should be 24, the actual line height is 28
3)   Chip should be spaced 8px apart when stack together horizontally or vertically
4)   For an delete able chip, when chip is highlighted, keypress on space and enter should able to delete this chip
5)   The Figma File Indicate there should 8px before and 8px after the close icon, the actual is 4px before and 8px after the close icon
6)  For Filter Chip, Leading Icon should not be allowed
7)  Chip Base should have 3px padding top and 3px padding bottom
8) Text should have 2px padding top, currently it's 1.6 padding bottom
9) Question, as figma indicated, the min height should 32px, now give 1px border top, + 3px chip padding top + 2px text padding top + 24 px line height + 3px chip padding bottom + 1 px border bottom= 34px. which one is correct?

**Comment by lizhuomeng71** [comment]
1) Chip should be spaced 8px apart when stack together horizontally or vertically
2) When Component is focused, the close icon should have hover state
image
3)  deletable="true" property should be default, and by the figma design, I think I close icon should always presented, not optional

**Comment by syedszeeshan** [comment]
@twjeffery 
Could you please confirm below?
Is deletable property supposed to be true by default?
Is the close icon always required or can we also have a basic chip without a close icon?

**Comment by twjeffery** [comment]
> @twjeffery Could you please confirm below? Is deletable property supposed to be true by default? Is the close icon always required or can we also have a basic chip without a close icon?

For the "Filter Chip" specifically, deletable property should be true by default. Close icon is always required for the filter chip as well.

**Comment by syedszeeshan** [comment]
@twjeffery As discussed, I am going to remove the deletable prop and adjust the code to alwasy show close icon.

**Comment by syedszeeshan** [comment]
> 1. Chip should be spaced 8px apart when stack together horizontally or vertically
> 2. When Component is focused, the close icon should have hover state
>    image
> 3. deletable="true" property should be default, and by the figma design, I think the close icon should always presented, not optional

@lizhuomeng71 I have fixed the 2 and 3. 

@lizhuomeng71 @twjeffery 
For #1, I think that 8px spacing should not be part of the component. The user can always specify the 8px margin as needed, you can see that in the test code I provided.

**Comment by lizhuomeng71** [comment]
This issue is verified on both react and angular

File: libs/web-components/src/components/filter-chip/FilterChip.svelte [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    line-height: var(--goa-line-height-2); /* 24px */
+    padding-top: 2px;
+    padding-bottom: var(
+      --font-valign-fix,
+      0
+    ); /* acumin font requires this to allow for vertical alignment  */
+    overflow-wrap: break-word;
+    word-wrap: break-word; /* if we need to support very old browsers */
+    /* hyphens: auto; */ /* This is not in Figma */
+    /* max-width: 288px; */ /* Should we set a max-width for long content? */
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/web-components/src/components/filter-chip/FilterChip.svelte:149]: @twjeffery do we want to apply any max-width for long content?

File: libs/web-components/src/components/filter-chip/FilterChip.svelte [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  // margin
+  export let mt: Spacing = null;
+  export let mr: Spacing = null;
+  export let mb: Spacing = null;
+  export let ml: Spacing = null;
+
+  // Props
+  export let error: string = "false";
+  export let content: string;
+  export let variant: ChipVariant;
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/web-components/src/components/filter-chip/FilterChip.svelte]: PR updated

File: libs/web-components/src/components/filter-chip/FilterChip.svelte [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+  export let testid: string = "";
+  export let ariaLabel: string = "";
+
+  // Private variables
+  let el: HTMLElement;
+  let _hovering: boolean = false;
+  let _focused: boolean = false;
+
+  // booleans
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/web-components/src/components/filter-chip/FilterChip.svelte:24]: PR updated

File: libs/web-components/src/components/filter-chip/FilterChip.svelte [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+  </div>
+  <goa-icon
+    class="delete-icon"
+    size="small"
+    theme={_error ? "filled" : _hovering || _focused ? "filled" : "outline"}
+    type="close-circle"
+    fillcolor={_error
+      ? "var(--goa-color-emergency-default)"
+      : "var(--goa-color-greyscale-700)"}
+    aria-hidden="true"
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/web-components/src/components/filter-chip/FilterChip.svelte]: PR updated

File: libs/web-components/src/components/filter-chip/FilterChip.svelte [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+<!-- Style -->
+<style>
+  .chip {
+    /* vertical-align: middle; */
+    display: inline-flex;
+    /* flex-direction: row; */
+    align-items: center;
+    background-color: var(--goa-color-greyscale-white);
+    border-radius: 1rem;
+    border-radius: var(--Border-Radius-Round, 99px);
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/web-components/src/components/filter-chip/FilterChip.svelte]: PR updated

File: libs/web-components/src/components/filter-chip/FilterChip.svelte (hunk 1 above)
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/web-components/src/components/filter-chip/FilterChip.svelte]: Confirmed with Thomas, max-width is not needed.
PR updated

File: libs/web-components/src/index.ts [hunk 6]
```diff
@@ -71,7 +72,3 @@ export { default as TextArea } from "./components/text-area/TextArea.svelte";
 export { default as ThreeColumnLayout } from "./layouts/three-column-layout/ThreeColumnLayout.svelte";
 export { default as Tooltip } from "./components/tooltip/Tooltip.svelte";
 export { default as TwoColumnLayout } from "./layouts/two-column-layout/TwoColumnLayout.svelte";
-
-// Experimental components
-export { default as ExperimentalFormStepper } from "./experimental/ExperimentalFormStepper.svelte";
-export { default as ExperimentalFormStep } from "./experimental/ExperimentalFormStep.svelte";
```
Comment (chrisolsen) [inline libs/web-components/src/index.ts:77]: Probably want to keep these

File: libs/react-components/src/lib/filter-chip/filter-chip.tsx [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+}: GoAFilterChipProps) => {
+  const el = useRef<HTMLElement>(null);
+  useEffect(() => {
+    if (!el.current) return;
+    if (!onClick) return;
+
+    const current = el.current;
+    const listener = () => {
+      onClick();
+    };
```
Comment (chrisolsen) [inline libs/react-components/src/lib/filter-chip/filter-chip.tsx:49]: Don't really need to wrap the function

File: libs/web-components/src/index.ts (hunk 6 above)
Comment (syedszeeshan) [reply libs/web-components/src/index.ts]: yes, PR updated.

File: libs/react-components/src/lib/filter-chip/filter-chip.tsx (hunk 7 above)
Comment (syedszeeshan) [reply libs/react-components/src/lib/filter-chip/filter-chip.tsx]: PR updated, removed the 'listener'.

File: libs/web-components/src/components/filter-chip/FilterChip.html-data.json [hunk 8]
```diff
@@ -0,0 +1,83 @@
+{
+  "name": "goa-filter-chip",
+  "required": "content",
+  "attributes": [
+    {
+      "name": "variant",
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/web-components/src/components/filter-chip/FilterChip.html-data.json]: PR updated.

File: libs/react-components/src/lib/filter-chip/filter-chip.tsx [hunk 9]
```diff
@@ -0,0 +1,69 @@
+import { useEffect, useRef } from "react";
+import { Margins } from "../../common/styling";
+
+export type GoAFilterChipTheme = "outline" | "filled" | "sharp";
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/react-components/src/lib/filter-chip/filter-chip.tsx]: PR updated.

**Review by chrisolsen — APPROVED** [review]

**Pull request merged**